src/ucp/rndv/rndv.c

```diff
@@ -397,6 +334,46 @@ static void ucp_rndv_complete_rma_get_zcopy(ucp_request_t *rndv_req,
     ucp_rndv_recv_req_complete(rreq, status);
 }
 
+static void ucp_rndv_complete_rma_put_zcopy(ucp_request_t *sreq, ucs_status_t status,
+                                            int is_frag_put)
+{
+    ucp_request_t *atp_req;
+
+    ucs_assertv(sreq->send.state.dt.offset <= sreq->send.length,
+                "sreq=%p offset=%zu length=%zu", sreq,
+                sreq->send.state.dt.offset, sreq->send.length);
+
+    /* complete send request after PUT completions of all fragments */
+    if (sreq->send.state.dt.offset != sreq->send.length) {
+        return;
+    }
```

Comment:
seems like status is ignored here? I guess we should save first failure status

Reply:
the status is updated in `uct_invoke_completion()` when PUT Zcopy is completed

Reply:
but at line 374 the request is completed with last fragment status (not completion status)
```
    ucp_request_complete_send(sreq, status);
```
what I'm missing?

Reply:
I've re-implemented it to make it more clear and less errorr prone
pls, take a look

> but at line 374 the request is completed with last fragment status (not completion status)
> 
> ```
>     ucp_request_complete_send(sreq, status);
> ```
> 
> what I'm missing?

we pass `self->status` to those function which is updated by the first failed status or `UCS_OK`

Reply:
> I've re-implemented it to make it more clear and less errorr prone
> pls, take a look
> 
> > but at line 374 the request is completed with last fragment status (not completion status)
> > ```
> >     ucp_request_complete_send(sreq, status);
> > ```
> > 
> > 
> > what I'm missing?
> 
> we pass `self->status` to those function which is updated by the first failed status or `UCS_OK`

but if we can return out from this function when `sreq->send.state.dt.offset != sreq->send.length` we lost the `self->status`, what if this request is on pending queue due to lack of resources for last fragment?

Reply:
But this function for a completion only, so it is called when in-progress PUT/GET Zcopy operation is completed, i.e. transport calls `uct_invoke_completion()` which updates the status

Reply:
So, we don't need to do any additional actions to save the status

Reply:
> But this function for a completion only, so it is called when in-progress PUT/GET Zcopy operation is completed, i.e. transport calls `uct_invoke_completion()` which updates the status

UCT completion status and UCP request completion status are not the same. are you sure that we use UCT completion to complete RNDV PUT/GET request from pending with correct UCT status in case of another request failure?
actually this case is not critical but shows how we can lost status.

can we do fast-forward in case of ERR since this is last posted completion (comp count == 0)?

Reply:
> UCT completion status and UCP request completion status are not the same

they are same in this case, we use `send.state.uct_comp`

> are you sure that we use UCT completion to complete RNDV PUT/GET request from pending with correct UCT status in case of another request failure?

this is a problem only for fragmented RNDV PUT
fixed by keeping send a request up to date